### PR TITLE
Add support for registration token management

### PR DESCRIPTION
**Pull request by govynnus**
This adds support for managing registration tokens using the admin API introduced by https://github.com/matrix-org/synapse/pull/10142 which implements [MSC3231 - Token Authenticated Registration](https://github.com/matrix-org/matrix-doc/pull/3231).

That synapse PR has been merged but isn't part of a release yet, so you probably won't want this until it is.
I'm happy to change names of commands/options/anything if you want.

### CODE REVIEW

**Comment by anoadragon453** [comment]
(https://matrix-org.github.io/synapse/develop/usage/administration/admin_api/registration_tokens.html now resolves by the way :)

**Comment by govynnus** [comment]
@JOJ0 It would be nice to display the expiry time in a human friendly format too (in `list` and `details`), but I'm not sure whether to do that in `api.py` or `cli/regtok.py`. I think it would just be adding another field to each dictionary.

**Comment by JOJ0** [comment]
Thanks for your patience. Been busy with music projects and work lately.

> I think it would just be adding another field to each dictionary.

You mean adding a new field in the results dict we get from the api response, ie additionally to the "expiry_time" field? Another option would be to just replace the value in expire_time with a human readable format. But in that case it should be configurable via another --option. Not sure if that's a good idea though. What do you think?

**Comment by govynnus** [comment]
Ooh, music projects :-)

> You mean adding a new field in the results dict we get from the api response, ie additionally to the "expiry_time" field

Yeah. I think the timestamp format should at least be available of it is wanted.  Maybe by default show human readable only and have an option that shows the timestamp instead? Because in normal use the timestamp will probably just be clutter.

> And regarding wheather to put it in api.py or cli frontend code, I think since the helper methods are in api.py and possibly api.py could be used standalone theoretically I feel it's ok that the api method can be asked te return human readable or unix timestamp.

Ok, thanks.

**Comment by JOJ0** [comment]
> > You mean adding a new field in the results dict we get from the api response, ie additionally to the "expiry_time" field
> 
> Yeah. I think the timestamp format should at least be available of it is wanted. Maybe by default show human readable only and have an option that shows the timestamp instead? Because in normal use the timestamp will probably just be clutter.

So you _do_ mean "replacing" the value of expiry_time and _not_ "adding a new additional field" ("expiry_date" or something)? And then have an option switch to control weather the value is a timestamp or an actual date? Did I understand correctly?

**Comment by govynnus** [comment]
Well to start with I did mean adding a new additional field,  but now I think it might be better to replace it and have an option switch.

**Comment by govynnus** [comment]
Does f0b6e0087180ec9bdde9ddfdfed54d16cca3ff09 look alright?

Registration token support has been released in Synapse 1.42.0, so the link to the docs works now.

**Comment by JOJ0** [comment]
Hi, thanks for letting me know that the feature is released now. Upgraded my Synapse and did some testing.
There is some issues with the datetime conversion. Have a look, yaml output is fine, in pprint it is literally showing the datetime object (the __repr__ I guess?):
```
(synadm) jojo@jum ~/git/govynnus_synadm (regtok) $ synadm -o yaml regtok list
registration_tokens:
- completed: 0
  expiry_time: 2021-09-30 00:00:00
  pending: 0
  token: lqvrdFcFXDTcQk3G
  uses_allowed: null

(synadm) jojo@jum ~/git/govynnus_synadm (regtok) $ synadm -o pprint regtok list
{'registration_tokens': [{'completed': 0,
                          'expiry_time': datetime.datetime(2021, 9, 30, 0, 0),
                          'pending': 0,
                          'token': 'lqvrdFcFXDTcQk3G',
                          'uses_allowed': None}]}
(synadm) jojo@jum ~/git/govynnus_synadm (regtok) $
(synadm) jojo@jum ~/git/govynnus_synadm (regtok) $ synadm -o human regtok list
token             uses_allowed      pending    completed  expiry_time
----------------  --------------  ---------  -----------  -------------------
lqvrdFcFXDTcQk3G                          0            0  2021-09-30 00:00:00
(synadm) jojo@jum ~/git/govynnus_synadm (regtok) $
```

And json output fails because the output renderer can't handle it:
```
(synadm) jojo@jum ~/git/govynnus_synadm (regtok) $ synadm -o json regtok list
Traceback (most recent call last):
  File "/home/jojo/.venvs/synadm/bin/synadm", line 11, in <module>
    load_entry_point('synadm', 'console_scripts', 'synadm')()
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/decorators.py", line 33, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/home/jojo/git/govynnus_synadm/synadm/cli/regtok.py", line 59, in regtok_list_cmd
    helper.output(regtoks)
  File "/home/jojo/git/govynnus_synadm/synadm/cli/__init__.py", line 176, in output
    click.echo(self.formatter(data))
  File "/usr/lib/python3.6/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib/python3.6/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python3.6/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib/python3.6/json/encoder.py", line 180, in default
    o.__class__.__name__)
TypeError: Object of type 'datetime' is not JSON serializable
(synadm) jojo@jum ~/git/govynnus_synadm (regtok) $
```

Not sure what's the best way to fix those issues. Let's think about it...

**Comment by govynnus** [comment]
Ah sorry, I only tested with the human output format. Does converting to a string manually as in 323c33d50c2a2a4cf38cd7503519f5450ac7895d look alright?

File: synadm/cli/regtok.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    generated by the server.""", show_default=True)
+@click.option(
+    "--length", "-l", type=int, default=16,
+    help="""The length of the randomly generated token if the token is not
+    specified.""", show_default=True)
+@click.option(
+    "--uses-allowed", "-u", type=int, default=None,
+    help="""The number of times the token can be used to complete a
+    registration before it becomes invalid.  [default: unlimited uses]""",
+    show_default=True)
```
Comment (JOJ0) [inline synadm/cli/regtok.py:92]: I think the show_default should be false here since you hardcoded it already.

File: synadm/cli/regtok.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+@click.option(
+    "--expiry-ts", "-t", type=int, default=None,
+    help="""The latest time the registration token is valid.
+    Given as the number of milliseconds since 1970-01-01 00:00:00 UTC.
+     [default: no expiry]""",
+    show_default=True)
```
Comment (JOJ0) [inline synadm/cli/regtok.py:98]: same here. hardcoded default already.

File: synadm/cli/regtok.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+@click.option(
+    "--expire-at", "-e", type=click.DateTime(), default=None,
+    help="""The latest time the registration token is valid.
+    See above for available date/time formats.  [default: no expiry]""",
+    show_default=True)
```
Comment (JOJ0) [inline synadm/cli/regtok.py:103]: and here.

**Comment by JOJ0** [comment]
Yes perfect, that probably is the best solution, have a string right away. Also thanks for renaming the variable to readable_expiry, I think that simple change adds to code readability a lot! :-)

Otherwise that all looks good and I think we are ready to merge. I found some little things about the show_default flags and postet inline.

Something else: I am wondering if there are places in synadm where I should adapt to your nice idea of having an option to choose for output format `"--datetime/--timestamp"`. I used the timestamp translation stuff in media and quarantine API's a lot, but I think the `list` commands over there don't spit out timestamps. So, if you come across anywhere else where timestamps are produced, give me a hint and I will also use your idea there to streamline user experience. Thanks!

File: synadm/cli/regtok.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+
+@regtok.command(name="update")
+@click.argument("token", type=str)
+@click.option(
+    "--uses-allowed", "-u", type=int, default=None,
+    help="""The number of times the token can be used to complete a
+    registration before it becomes invalid. Use -1 for an unlimited
+    number of uses.  [default: unchanged]""",
+    show_default=True)
```
Comment (JOJ0) [inline synadm/cli/regtok.py:124]: or maybe you just get rid of the show_default flag everywhere you already hardcoded it? not sure but actually the default is False anyway.

File: synadm/cli/regtok.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    "--expiry-ts", "-t", type=int, default=None,
+    help="""The latest time the registration token is valid.
+    Given as the number of milliseconds since 1970-01-01 00:00:00 UTC.
+    Use -1 for no expiration.  [default: unchanged]""",
+    show_default=True)
+@click.option(
+    "--expire-at", "-e", type=click.DateTime(), default=None,
+    help="""The latest time the registration token is valid.
+    See above for available date/time formats.  [default: unchanged]""",
+    show_default=True)
```
Comment (JOJ0) [inline synadm/cli/regtok.py:135]: show_default

**Comment by govynnus** [comment]
I've removed `show_default` where the default is already hardcoded in the help text. I think I must have been copy/pasting :upside_down_face: 

If I do see any timestamps being output I'll let you know.

**Comment by JOJ0** [comment]
Ah one last thing, you used -t option twice in regtok new command ;-)

**Comment by JOJ0** [comment]
Found one more. This error should be catched/prevented. The output mode doesn't matter:
```
(synadm) jojo@jum ~/git/govynnus_synadm (regtok) $ synadm -o pprint regtok details nonexistenttoken
WARNING Synapse returned status code 404
Traceback (most recent call last):
  File "/home/jojo/.venvs/synadm/bin/synadm", line 11, in <module>
    load_entry_point('synadm', 'console_scripts', 'synadm')()
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/jojo/.venvs/synadm/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/decorators.py", line 33, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/home/jojo/git/govynnus_synadm/synadm/cli/regtok.py", line 72, in regtok_details_cmd
    regtok = helper.api.regtok_details(token, datetime)
  File "/home/jojo/git/govynnus_synadm/synadm/api.py", line 705, in regtok_details
    if result["expiry_time"] is not None:
KeyError: 'expiry_time'
(synadm) jojo@jum ~/git/govynnus_synadm (regtok) $
```

I guess you could quickly fix that KeyError with a result.get("expiry_time")?

**Comment by JOJ0** [comment]
Hi, I'll merge this now and quickly fix these little things in the master. Thank you so much for this very complete PR, your thoroughness and good communication! May I invite you to #synadm:peek-a-boo.at, we often share opinions about new features and how they could/should be designed. I am sure your presence and thoughts would add some to the community. All the best. Jojo

**Comment by JOJ0** [comment]
Oh, how nice, I spotted your summary page about the feature. Great contribution to Matrix in general. Thanks for that!!! https://calcuode.com/matrix-gsoc/2021-08-22_final-report.html
You could mention synadm as supporting it entirely now if you want. Thanks again for the PR!

**Comment by govynnus** [comment]
Hi, sorry I didn't get round to this, been busy moving to university the past ~week. I am planning to write another blog post now that the synapse part is released and synadm support exists :-)

Thanks for your help with this  and thanks for synadm, probably see you around in #synadm:peek-a-boo.at

**Pull request merged**